query/src/main/java/io/keen/client/java/KeenQueryClient.java

```diff
@@ -0,0 +1,704 @@
+        Map<AbsoluteTimeframe, QueryResult> intervalResult = new HashMap<AbsoluteTimeframe, QueryResult>();
+
+        for (Object child : intervals) {
+            if (child instanceof HashMap) {
+                HashMap<String, Object> inputMap = (HashMap<String, Object>) child;
+                // If this is an interval, it should have keys "timeframe" and "value"
+                if (inputMap.containsKey(KeenQueryConstants.TIMEFRAME) && (inputMap.containsKey(KeenQueryConstants.VALUE))) {
+                    AbsoluteTimeframe absoluteTimeframe = null;
+                    Object timeframe = inputMap.get(KeenQueryConstants.TIMEFRAME);
+                    if (timeframe instanceof HashMap) {
+                        HashMap<String, String> hashTimeframe = (HashMap<String, String>) timeframe;
+                        String start = hashTimeframe.get(KeenQueryConstants.START);
+                        String end = hashTimeframe.get(KeenQueryConstants.END);
+                        absoluteTimeframe = new AbsoluteTimeframe(start, end);
+                    }
+
+                    Object value = inputMap.get(KeenQueryConstants.VALUE);
+                    QueryResult queryResultValue = constructQueryResult(value, isGroupBy, false);
+
+                    intervalResult.put(absoluteTimeframe, queryResultValue);
+                } else {
+                    throw new IllegalStateException();
+                }
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+
+        return new IntervalResult(intervalResult);
+    }
+
+    private static GroupByResult constructGroupByResult(ArrayList<Object> groups) {
+        Map<Group, QueryResult> groupByResult = new HashMap<Group, QueryResult>();
+
+        for (Object child : groups) {
+            if (child instanceof HashMap) {
+                HashMap<String, Object> inputMap = (HashMap<String, Object>) child;
+
+                // If this is a GroupByResult, it should have key "result", along with properties to group by.
+                if (inputMap.containsKey(KeenQueryConstants.RESULT)) {
+                    QueryResult result = null;
+                    HashMap<String, Object> properties = new HashMap<String, Object>();
+                    for (String key : inputMap.keySet()) {
+                        if (key.equals(KeenQueryConstants.RESULT)) {
+                            // there should not be intervals nested inside GroupByResult's; only
+                            // the other way around.
+                            result = constructQueryResult(inputMap.get(key), false, false);
+                        } else {
+                            properties.put(key, inputMap.get(key));
+                        }
+                    }
+
+                    Group groupBy = new Group(properties);
+                    groupByResult.put(groupBy, result);
+                } else {
+                    throw new IllegalStateException();
```

Comment:
Suggestion: Add messages to all of these `IllegalStateExceptions`. With no message, debugging would be a bit of a pain. This applies to many places throughout the PR.